### PR TITLE
XW-904 | Replace $.ajax with ember-ajax

### PR DESCRIPTION
**Pull request by hakubo**
## Links

* https://wikia-inc.atlassian.net/browse/XW-904
* https://github.com/ember-cli/ember-ajax
* https://wikia-inc.atlassian.net/wiki/display/MER/2016/04/22/ember-ajax+-+an+ajax+lib+that+also+works+with+FastBoot

## Description

This PR prepares us to use Ember Fastboot, but before we get there we need to have a way to call APIs with the same code both on client and server side. Ember ajax has a basic support for that and they are committed to improve that part of the lib.

## Reviewers

@kvas-damian @bkoval @Wikia/x-wing

### CODE REVIEW

**Comment by hakubo** [comment]
There are couple of issues I have found that I'll fix hopefully on Friday
Feel free to code review that though

Thanks!

**Comment by kvas-damian** [comment]
@hakubo `ajax-call` that we use for discussions is dumb wrapper on `$.ajax` but we put there some default options that we didn't want to repeat in every place. I believe it's valuable for keeping this helper in place - using `ember-ajax` and probably renaming to make this helper discussion specific.

**Comment by rogatty** [comment]
@kvas-damian what exactly do you believe is valuable in `utils/ajax-call` and is now repeated through the code? From what I see `ember-ajax` does the similar thing - sets some defaults and removes the need for `Ember.RSVP` wrappers.

**Comment by kvas-damian** [comment]
@rogatty 
`contentType: 'application/json',
dataType: 'json',
xhrFields: {
	withCredentials: true
}`

**Comment by rogatty** [comment]
See https://github.com/Wikia/mercury/pull/2378/files#diff-3010be3e471d7fd9d1249aab192b1135R6
```
$.ajaxSetup({
	cache: true,
	contentType: 'application/json; charset=utf-8',
	xhrFields: {
		withCredentials: true
	}
});
```
:point_up: these are now defaults for all ajax requests.

`dataType: 'json'` is the default in `ember-ajax` itself.

**Comment by kvas-damian** [comment]
Ohh didn't see that :bowtie: . Makes sense now ;)

File: front/common/baseline/build-url.js [hunk 1]
```diff
@@ -132,8 +132,8 @@ if (typeof window.M === 'undefined') {
 	 * @param {Object} [query={}]
```
Comment (tomasznapieralski) [inline front/common/baseline/build-url.js]: this is no longer valid

File: front/main/app/components/article-content.js [hunk 2]
```diff
@@ -202,32 +202,38 @@ export default Ember.Component.extend(
 			return '';
 		},
 
+		handleAttrsContext(attrs) {
```
Comment (tomasznapieralski) [inline front/main/app/components/article-content.js]: please add jsdoc

File: front/main/app/components/article-content.js [hunk 3]
```diff
@@ -202,32 +202,38 @@ export default Ember.Component.extend(
...
+				 * Perfectly this should be handled somewhere inside infobox-related logic.
+				 * For now this solution is enough
+				 * - it works the same way as on wikis without SEO friendly images.
+				 * It works on wikis without SEO friendly images because there was a bug
+				 * - video was treated as an image and we don't show titles below images.
+				 */
+				if (attrs.context === 'infobox' && attrs.type === 'video') {
+					attrs.showTitle = false;
+				}
+				attrs.mediaContext = attrs.context;
```
Comment (tomasznapieralski) [inline front/main/app/components/article-content.js]: please provide space above

File: front/main/app/components/article-content.js [hunk 4]
```diff
@@ -202,32 +202,38 @@ export default Ember.Component.extend(
...
-						 * It works on wikis without SEO friendly images because there was a bug
-						 * - video was treated as an image and we don't show titles below images.
-						 */
-						if (attrs.context === 'infobox' && attrs.type === 'video') {
-							attrs.showTitle = false;
-						}
-						attrs.mediaContext = attrs.context;
-						delete attrs.context;
-					}
+					attrs = this.handleAttrsContext(attrs);
```
Comment (tomasznapieralski) [inline front/main/app/components/article-content.js]: Did you thought about this:
`this.handleAttrsContext(Ember.$.extend(attrs, media[attrs.ref]))`

File: front/main/app/mixins/discussion-moderation-controller.js [hunk 5]
```diff
@@ -1,6 +1,8 @@
 import Ember from 'ember';
 
-export default Ember.Mixin.create({
+const {Mixin} = Ember;
```
Comment (tomasznapieralski) [inline front/main/app/mixins/discussion-moderation-controller.js]: you could destruct Ember object directly in import section

File: front/main/app/mixins/discussion-moderation-model.js [hunk 6]
```diff
@@ -1,25 +1,24 @@
 import Ember from 'ember';
-import ajaxCall from '../utils/ajax-call';
+import request from 'ember-ajax/request';
 
-export default Ember.Mixin.create({
+const {Mixin} = Ember;
```
Comment (tomasznapieralski) [inline front/main/app/mixins/discussion-moderation-model.js]: you could destruct Ember object directly in import section

File: front/main/app/models/article-add-photo.js [hunk 7]
```diff
@@ -64,38 +65,30 @@ ArticleAddPhotoModel.reopenClass(
...
-						}
-					},
-					error: (err) => reject(err)
-				});
+		editContent(data) {
+			return request(M.buildUrl({path: '/api.php'}), {
+				method: 'POST',
+				data,
+			}).then((resp) => {
+				if (resp && resp.edit && resp.edit.result === 'Success') {
```
Comment (bognix) [inline front/main/app/models/article-add-photo.js]: how about simplifying to `Ember.get(resp, 'edit.result') === 'Success)` ?

File: front/main/app/mixins/discussion-moderation-route.js [hunk 8]
```diff
@@ -1,7 +1,9 @@
 import Ember from 'ember';
 
-export default Ember.Mixin.create({
-	modalDialogService: Ember.inject.service('modal-dialog'),
+const {Mixin, inject, set} = Ember;
```
Comment (tomasznapieralski) [inline front/main/app/mixins/discussion-moderation-route.js]: you could destruct Ember object directly in import section

File: front/main/app/mixins/discussion-moderation-route.js [hunk 9]
```diff
@@ -1,7 +1,9 @@
...
+
+export default Mixin.create({
+	modalDialogService: inject.service('modal-dialog'),
```
Comment (tomasznapieralski) [inline front/main/app/mixins/discussion-moderation-route.js]: can you use:
`modalDialog: inject.service()`?

File: front/main/app/models/article-add-photo.js [hunk 10]
```diff
@@ -136,30 +125,23 @@ ArticleAddPhotoModel.reopenClass(
...
-								} else if (resp && resp.error) {
-									reject(resp.error.code);
-								} else {
-									reject();
-								}
-							},
-							error: (err) => reject(err)
-						});
+						},
+					}).then((resp) => {
```
Comment (tomasznapieralski) [inline front/main/app/models/article-add-photo.js]: can you use `response`, please?

File: front/main/app/models/article-add-photo.js [hunk 11]
```diff
@@ -178,28 +160,21 @@ ArticleAddPhotoModel.reopenClass(
...
-							error: (err) => reject(err)
-						});
+
+					return request(M.buildUrl({path: '/api.php'}), {
+						method: 'POST',
+						data: formData,
+						cache: false,
+						contentType: false,
+						processData: false,
+					}).then((resp) => {
```
Comment (tomasznapieralski) [inline front/main/app/models/article-add-photo.js]: can you use 'response` please?

File: front/main/app/models/article-diff.js [hunk 12]
```diff
@@ -31,37 +32,33 @@ const ArticleDiffModel = Ember.Object.extend({
...
+					data: {
+						action: 'edit',
+						summary,
+						title: this.title,
+						undo: this.newId,
+						undoafter: this.oldId,
+						token,
+						format: 'json'
+					},
+				}).then((resp) => {
```
Comment (tomasznapieralski) [inline front/main/app/models/article-diff.js]: can you use `response` please?

File: front/main/app/models/article-diff.js [hunk 13]
```diff
@@ -31,37 +32,33 @@ const ArticleDiffModel = Ember.Object.extend({
...
-							format: 'json'
-						},
-						dataType: 'json',
-						method: 'POST',
-						success: (resp) => {
-							if (resp && resp.edit && resp.edit.result === 'Success') {
-								resolve();
-							} else if (resp && resp.error) {
-								const errorMsg = resp.error.code === 'undofailure' ? 'main.undo-failure' : 'main.error';
+		return getEditToken(this.title)
```
Comment (tomasznapieralski) [inline front/main/app/models/article-diff.js]: can you use `this.get('title')`

File: front/main/app/models/article-diff.js [hunk 14]
```diff
@@ -31,37 +32,33 @@ const ArticleDiffModel = Ember.Object.extend({
...
+			.then((token) => {
+				request(M.buildUrl({path: '/api.php'}), {
+					method: 'POST',
+					contentType: 'application/x-www-form-urlencoded; charset=UTF-8',
+					data: {
+						action: 'edit',
+						summary,
+						title: this.title,
```
Comment (tomasznapieralski) [inline front/main/app/models/article-diff.js]: can you use `this.get('title')`?

File: front/main/app/models/article-diff.js [hunk 15]
```diff
@@ -31,37 +32,33 @@ const ArticleDiffModel = Ember.Object.extend({
...
+						undo: this.newId,
```
Comment (tomasznapieralski) [inline front/main/app/models/article-diff.js]: can you use `this.get('newId')`?

File: front/main/app/models/article-diff.js [hunk 16]
```diff
@@ -31,37 +32,33 @@ const ArticleDiffModel = Ember.Object.extend({
...
+						undoafter: this.oldId,
```
Comment (tomasznapieralski) [inline front/main/app/models/article-diff.js]: can you use `this.get('oldIt')`?

File: front/main/app/models/article-diff.js [hunk 17]
```diff
@@ -73,51 +70,46 @@ ArticleDiffModel.reopenClass({
...
-				let modelInstance = null,
-					lengthChange = revision.size;
+			let modelInstance = null,
+				lengthChange = revision.size;
 
-				if (!Ember.isNone(oldRevision)) {
-					lengthChange = lengthChange - oldRevision.size;
-				}
+			if (!Ember.isNone(oldRevision)) {
+				lengthChange = lengthChange - oldRevision.size;
```
Comment (tomasznapieralski) [inline front/main/app/models/article-diff.js]: you can use shortcut `-=`

File: front/main/app/models/article-edit.js [hunk 18]
```diff
@@ -19,34 +20,29 @@ ArticleEditModel.reopenClass(
...
+						contentType: 'application/x-www-form-urlencoded; charset=UTF-8',
+						data: {
+							action: 'edit',
+							title: model.title,
+							section: model.sectionIndex,
+							text: model.content,
+							token,
+							format: 'json'
+						},
+					}).then((resp) => {
```
Comment (tomasznapieralski) [inline front/main/app/models/article-edit.js]: can you use 'response`?

**Comment by bognix** [comment]
I like it! LGTM

File: front/main/app/models/article-edit.js [hunk 19]
```diff
@@ -55,45 +51,40 @@ ArticleEditModel.reopenClass(
...
+				data: {
+					action: 'query',
+					prop: 'revisions',
+					// FIXME: It should be possible to pass props as an array
+					rvprop: 'content|timestamp',
+					titles: title,
+					rvsection: sectionIndex,
+					format: 'json'
+				}
+			}).then((resp) => {
```
Comment (tomasznapieralski) [inline front/main/app/models/article-edit.js]: can you use `response`?

File: front/main/app/models/article-edit.js [hunk 20]
```diff
@@ -55,45 +51,40 @@ ArticleEditModel.reopenClass(
...
-							content: revision['*'],
-							originalContent: revision['*'],
-							timestamp: revision.timestamp
-						}));
-					} else {
-						reject();
-					}
-				}).fail((err) => reject(err));
+				if (pages) {
+					// FIXME: MediaWiki API, seriously?
```
Comment (tomasznapieralski) [inline front/main/app/models/article-edit.js]: do we need this comment?

File: front/main/app/mixins/discussion-moderation-controller.js (hunk 5 above)
Comment (hakubo) [inline front/main/app/mixins/discussion-moderation-controller.js]: unfortunatelly I can't as Ember does not have proper exports yet

File: front/main/app/models/article-diff.js (hunk 17 above)
Comment (hakubo) [inline front/main/app/models/article-diff.js]: I don't think it adds to readability here. I'll leave it as is now

File: front/main/app/models/article-edit.js (hunk 20 above)
Comment (hakubo) [inline front/main/app/models/article-edit.js]: don't know - out of scope of this PR anyway now I think :)

File: front/main/app/mixins/discussion-moderation-controller.js (hunk 5 above)
Comment (tomasznapieralski) [inline front/main/app/mixins/discussion-moderation-controller.js]: 👍

File: front/main/app/models/article-diff.js (hunk 17 above)
Comment (tomasznapieralski) [inline front/main/app/models/article-diff.js]: 👍

File: front/main/app/models/article-edit.js (hunk 20 above)
Comment (tomasznapieralski) [inline front/main/app/models/article-edit.js]: 👍

**Comment by tomasznapieralski** [comment]
LGTM

File: front/common/tests/unit/baseline/build-url.js [hunk 21]
```diff
@@ -213,29 +213,19 @@ QUnit.module('M.buildUrl helper function (loaded with baseline)', function (hook
...
-				},
-				expectedOutput: 'https://services.wikia.com/discussion/147/threads/2?' +
-				'responseGroup=full&sortDirection=descending&sortKey=creation_date&limit=3'
+				expectedOutput: 'https://services.wikia.com/discussion/147/threads/2'
 			},
 			{
 				path: '',
-				query: {action: '&=/'},
-				expectedOutput: 'https://services.wikia.com/discussion?action=%26%3D%2F'
+				expectedOutput: 'https://services.wikia.com/discussion'
```
Comment (rogatty) [inline front/common/tests/unit/baseline/build-url.js]: These two cases no longer make sense, I'd remove them.

File: front/main/app/models/article-diff.js [hunk 22]
```diff
@@ -31,37 +32,33 @@ const ArticleDiffModel = Ember.Object.extend({
...
-						success: (resp) => {
-							if (resp && resp.edit && resp.edit.result === 'Success') {
-								resolve();
-							} else if (resp && resp.error) {
-								const errorMsg = resp.error.code === 'undofailure' ? 'main.undo-failure' : 'main.error';
+		return getEditToken(this.get('title'))
+			.then((token) => {
+				request(M.buildUrl({path: '/api.php'}), {
+					method: 'POST',
+					contentType: 'application/x-www-form-urlencoded; charset=UTF-8',
```
Comment (rogatty) [inline front/main/app/models/article-diff.js]: @hakubo I wonder if we should abstract this string somewhere? Util maybe? It won't be easier to use than copy-paste but it'll be harder to make a mistake.

File: front/main/app/services/revision-upvotes.js [hunk 23]
```diff
@@ -74,32 +75,28 @@ export default Ember.Service.extend({
...
+		return getEditToken(title)
+			.then((token) => {
+				return this.get('ajax').post(M.buildUrl({
+					path: '/wikia.php',
+					query: {
+						controller: 'RevisionUpvotesApiController',
+						method: 'addUpvote'
+					}
+				}), {
+					contentType: 'application/x-www-form-urlencoded; charset=utf-8',
```
Comment (rogatty) [inline front/main/app/services/revision-upvotes.js]: :speak_no_evil:

File: front/main/app/services/revision-upvotes.js [hunk 24]
```diff
@@ -112,32 +109,28 @@ export default Ember.Service.extend({
...
+		return getEditToken(title)
+			.then((token) => {
+				return this.get('ajax').post(M.buildUrl({
+					path: '/wikia.php',
+					query: {
+						controller: 'RevisionUpvotesApiController',
+						method: 'removeUpvote'
+					}
+				}), {
+					contentType: 'application/x-www-form-urlencoded; charset=UTF-8',
```
Comment (rogatty) [inline front/main/app/services/revision-upvotes.js]: :see_no_evil:

File: front/main/tests/unit/utils/wiki-handlers/wiki-page-test.js [hunk 25]
```diff
@@ -4,7 +4,7 @@ import {getModelForNamespace} from 'main/utils/wiki-handlers/wiki-page';
 import Ember from 'ember';
 import sinon from 'sinon';
 
-module('Unit | Utils | wiki-handlers/wiki-page', (hooks) => {
+module('Unit | Utility | wiki-handlers/wiki-page', (hooks) => {
```
Comment (rogatty) [inline front/main/tests/unit/utils/wiki-handlers/wiki-page-test.js]: Why would you? It's `Utils` in every other util test.

File: front/main/tests/unit/utils/content-type-test.js [hunk 26]
```diff
@@ -0,0 +1,10 @@
+import {form, json} from 'main/utils/content-type';
+import {module, test} from 'qunit';
+
+module('Unit | Utility | content type');
+
+// Replace this with your real tests.
```
Comment (rogatty) [inline front/main/tests/unit/utils/content-type-test.js]: I don't think we should keep this, they're very real!

**Comment by rogatty** [comment]
I left some minor comments but generally LGTM.

File: front/main/tests/unit/utils/wiki-handlers/wiki-page-test.js (hunk 25 above)
Comment (hakubo) [inline front/main/tests/unit/utils/wiki-handlers/wiki-page-test.js]: This is what ember g util - generates
I can rename it in all files

**Pull request merged**